src/index.ts

```diff
@@ -0,0 +1,522 @@
+
+interface IamdeeDefineFunction {
+  (factory: IamdeeDefineFactory): void;
+  (dependencies: string[], factory: IamdeeDefineFactory): void;
+  (id: string, dependencies: string[], factory: IamdeeDefineFactory): void;
+  (id: string, factory: IamdeeDefineFactory): void;
+  <T extends IamdeeDefineArgs>(...args: T): void;
+  amd: {};
+}
+
+interface Window {
+  define: IamdeeDefineFunction;
+  require: IamdeeRequireFunction;
+  requirejs: IamdeeRequireFunction;
+}
+
+(function(window, undefined) {
+  interface ArrayMap<TKey, TValue> {
+    get: (key: TKey) => TValue | undefined;
+    set: (key: TKey, value: TValue) => void;
+  }
+
+  function map<TKey, TValue>(): ArrayMap<TKey, TValue> {
+    const keys: TKey[] = [];
+    const values: TValue[] = [];
+    return {
+      get(key) {
+        return values[keys.indexOf(key)];
+      },
+      set(key, value) {
+        const index = keys.indexOf(key);
+        if (index == -1) {
+          keys.push(key);
+          values.push(value);
+        } else {
+          values[index] = value;
+        }
+      }
+    };
+  }
+
+  const enum ModuleState {
+    DEFINED = 0,
+    NETWORK_LOADING = 1,
+    WAITING_FOR_DEPENDENCIES = 2,
+    INITIALIZED = 3,
+    ERROR = 4
+  }
+
+  type ModuleId =
+    | string & { readonly __tag: unique symbol }
+    | "require"
+    | "exports"
+    | "module";
+  type ModulePath = string & { readonly __tag: unique symbol };
+  type SourceUrl = string & { readonly __tag: unique symbol };
+  type RequestId = string & { readonly __tag: unique symbol };
+
+  interface ModuleCallback {
+    (module: Module): void;
+  }
+
+  interface DefinedModule {
+    id: ModuleId;
+    requestId?: RequestId;
+    state: ModuleState.DEFINED;
+    load: (requestId: RequestId) => void;
+    callbacks: ModuleCallback[];
+  }
+
+  interface NetworkLoadingModule {
+    id: ModuleId;
+    requestId: RequestId;
+    state: ModuleState.NETWORK_LOADING;
+    callbacks: ModuleCallback[];
+  }
+
+  interface WaitingForDependenciesModule {
+    id: ModuleId;
+    requestId: RequestId;
+    state: ModuleState.WAITING_FOR_DEPENDENCIES;
+    callbacks: ModuleCallback[];
+    exports: {};
+  }
+
+  interface InitializedModule {
+    id: ModuleId;
+    state: ModuleState.INITIALIZED;
+    exports: unknown;
+  }
+
+  interface ErrorModule {
+    id: ModuleId;
+    state: ModuleState.ERROR;
+    error: Error;
+  }
+
+  type Module =
+    | DefinedModule
+    | NetworkLoadingModule
+    | WaitingForDependenciesModule
+    | InitializedModule
+    | ErrorModule;
+
+  const moduleMap: ArrayMap<ModuleId, Module> = map();
+  const elementMap: ArrayMap<Element, ModuleId> = map();
+  function noop() {}
+
+  function panic(message: string) {
+    throw Error(message);
+  }
+
+  const doc = document;
+  let baseUrl = "./";
+  let onNodeCreated: IamdeeNodeCreatedCallback = noop;
+
+  function isAnonymousDefine(
+    args: IamdeeDefineArgs
+  ): args is IamdeeAnonymousDefineArgs {
+    return typeof args[0] != "string";
```

Comment:
shouldn't this be `typeof args[0] === "function"`

Reply:
Nope, because all of these anonymous define calls are supported:

```js
define(["some-rep"], function (somedep) { ... });
define(function (somedep) { ... });
define({ inlineObjProp: 42 });
```

Named define is when first arg is a string, e.g.:

```js
define("jquery", function () { ... });
```